### PR TITLE
feat(marketplace): add reviews and ratings system

### PR DESCRIPTION
**Pull request by BenjaminPrice**
## What does this PR do?

Adds a complete review and rating system to the marketplace. Users can submit star ratings (1-5) with optional text reviews via GitHub auth. Plugin owners can reply to reviews. Ratings are denormalized on the plugins table and shown in plugin detail when 3+ reviews exist.

## Type of change

- [ ] Bug fix
- [x] Feature (requires [approved Discussion](https://github.com/emdash-cms/emdash/discussions/categories/ideas))
- [ ] Refactor (no behavior change)
- [ ] Documentation
- [ ] Performance improvement
- [ ] Tests
- [ ] Chore (dependencies, CI, tooling)

## Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/emdash-cms/emdash/blob/main/CONTRIBUTING.md)
- [x] `pnpm typecheck` passes
- [x] `pnpm --silent lint:json | jq '.diagnostics | length'` returns 0 (1 pre-existing warning in router.tsx)
- [x] `pnpm test` passes (3 pre-existing failures in auth tests, 0 new failures)
- [x] `pnpm format` has been run
- [x] I have added/updated tests for my changes (if applicable)
- [x] I have added a [changeset](https://github.com/emdash-cms/emdash/blob/main/CONTRIBUTING.md#changesets) (if this PR changes a published package)
- [x] New features link to an approved Discussion: https://github.com/emdash-cms/emdash/discussions/296

## AI-generated code disclosure

- [x] This PR includes AI-generated code

### Schema (`packages/marketplace/src/db/schema.sql`)
- `reviews` table: id, plugin_id, author_id, rating (CHECK 1-5), body, publisher_reply, replied_at, timestamps, UNIQUE(plugin_id, author_id)
- Indexes: idx_reviews_plugin, idx_reviews_author
- `rating_avg REAL` and `rating_count INTEGER` columns on plugins table

### Review routes (`packages/marketplace/src/routes/reviews.ts`, new file)
- `GET /plugins/:id/reviews` -- paginated review list (public, cursor-based)
- `POST /plugins/:id/reviews` -- submit review (auth + rate limited, 1/min/author)
- `PUT /plugins/:id/reviews/:reviewId` -- update own review
- `DELETE /plugins/:id/reviews/:reviewId` -- delete own review
- `POST /plugins/:id/reviews/:reviewId/reply` -- publisher reply (plugin owner only)
- JWT auth middleware (same pattern as author routes, no seed token)

### Review queries (`packages/marketplace/src/db/queries.ts`)
- `getPluginReviews()` -- paginated with author JOIN
- `createReview()` -- with HTML sanitization + rating recalculation
- `updateReview()` -- own review only + rating recalculation
- `deleteReview()` -- own review only + rating recalculation
- `addPublisherReply()` -- verifies plugin ownership via JOIN
- `recalculatePluginRating()` -- atomic SQL UPDATE with AVG/COUNT subquery
- `sanitizeReviewText()` -- strips HTML tags, plain text only

### Plugin detail (`packages/marketplace/src/routes/public.ts`)
- Rating summary included when `rating_count >= 3`: `{ average, count }`
- Returns `null` when fewer than 3 reviews (prevents "5.0 (1 review)" trust issues)
- `rating` added as a sort option for search

### Design decisions
- **Plain text only** (no markdown) to minimize XSS attack surface
- **HTML stripped on write** via regex, not just on render
- **Rating hidden below 3 reviews** to prevent misleading scores
- **Atomic denormalization** via SQL subquery, no read-modify-write TOCTOU
- **In-memory rate limiting** per author (1 review/minute), with cleanup at 1000 entries

### Tests (15 new in `packages/marketplace/tests/reviews.test.ts`)
- Review CRUD: create, duplicate rejection (409), list, update, delete
- Auth: rejects unauthenticated (401), validates rating range (400)
- Edge cases: nonexistent plugin (404), HTML sanitization (strips `<script>` tags)
- Rating denormalization: avg/count update on create, recalculation on delete
- Rating display: hidden when < 3 reviews, shown when >= 3 with correct average
- Publisher reply: owner can reply (200), non-owner rejected (404)

### Migration note for production D1
```sql
ALTER TABLE plugins ADD COLUMN rating_avg REAL NOT NULL DEFAULT 0;
ALTER TABLE plugins ADD COLUMN rating_count INTEGER NOT NULL DEFAULT 0;

CREATE TABLE IF NOT EXISTS reviews (
  id TEXT PRIMARY KEY,
  plugin_id TEXT NOT NULL REFERENCES plugins(id),
  author_id TEXT NOT NULL REFERENCES authors(id),
  rating INTEGER NOT NULL CHECK(rating BETWEEN 1 AND 5),
  body TEXT,
  publisher_reply TEXT,
  replied_at TEXT,
  created_at TEXT NOT NULL DEFAULT (datetime('now')),
  updated_at TEXT NOT NULL DEFAULT (datetime('now')),
  UNIQUE(plugin_id, author_id)
);
CREATE INDEX IF NOT EXISTS idx_reviews_plugin ON reviews(plugin_id);
CREATE INDEX IF NOT EXISTS idx_reviews_author ON reviews(author_id);
```

### CODE REVIEW

File: packages/marketplace/src/routes/reviews.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+const updateReviewSchema = z.object({
+	rating: z.number().int().min(1).max(5).optional(),
+	body: z.string().max(5000).optional(),
+});
+
+reviewRoutes.put("/plugins/:id/reviews/:reviewId", async (c) => {
+	const reviewId = c.req.param("reviewId");
+	const author = c.get("author");
+
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/routes/reviews.ts:207]: Valid. updateReview() now takes an optional pluginId and verifies the review belongs to that plugin. Fixed in 82915d0.

File: packages/marketplace/src/routes/reviews.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	}
+});
+
+// ── DELETE /plugins/:id/reviews/:reviewId — Delete own review ──
+
+reviewRoutes.delete("/plugins/:id/reviews/:reviewId", async (c) => {
+	const reviewId = c.req.param("reviewId");
+	const author = c.get("author");
+
+	try {
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/routes/reviews.ts:242]: Valid. deleteReview() now takes an optional pluginId and verifies the review belongs to that plugin. Fixed in 82915d0.

File: packages/marketplace/src/routes/reviews.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+// ── POST /plugins/:id/reviews/:reviewId/reply — Publisher reply ─
+
+const replySchema = z.object({
+	body: z.string().min(1).max(2000),
+});
+
+reviewRoutes.post("/plugins/:id/reviews/:reviewId/reply", async (c) => {
+	const reviewId = c.req.param("reviewId");
+	const author = c.get("author");
+
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/routes/reviews.ts:263]: Valid. addPublisherReply() now takes an optional pluginId and verifies the review belongs to that plugin. Fixed in 82915d0.

File: packages/marketplace/src/db/queries.ts [hunk 4]
```diff
@@ -199,6 +203,190 @@ export async function getPluginVersion(
...
+			JOIN plugins p ON p.id = r.plugin_id
+			WHERE r.id = ? AND p.author_id = ?`,
+		)
+		.bind(reviewId, pluginOwnerId)
+		.first<{ plugin_id: string }>();
+
+	if (!review) return null;
+
+	await db
+		.prepare("UPDATE reviews SET publisher_reply = ?, replied_at = datetime('now') WHERE id = ?")
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/db/queries.ts]: Valid. Added updated_at = datetime('now') to the publisher reply UPDATE statement. Fixed in 82915d0.

File: packages/marketplace/src/routes/reviews.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	if (last && now - last < REVIEW_RATE_LIMIT_MS) {
+		return false;
+	}
+	recentReviews.set(authorId, now);
+	// Clean old entries periodically
+	if (recentReviews.size > 1000) {
+		for (const [key, ts] of recentReviews) {
+			if (now - ts > REVIEW_RATE_LIMIT_MS) recentReviews.delete(key);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/routes/reviews.ts:74]: Valid. Refactored to bounded rate limiter with MAX_TRACKED_AUTHORS=1000, pruning expired entries on every check and evicting oldest when over cap. Fixed in 82915d0.

File: packages/marketplace/src/routes/reviews.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+			return c.json({ error: "Validation error", details: err.errors }, 400);
+		}
+		return c.json({ error: "Invalid JSON" }, 400);
+	}
+
+	if (!checkRateLimit(author.id)) {
+		return c.json({ error: "Too many reviews. Please wait before submitting another." }, 429);
+	}
+
+	try {
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/routes/reviews.ts]: Valid. Split into isRateLimited() (check only) and recordRateLimit() (called after successful DB insert). Failed requests no longer consume the quota. Fixed in 82915d0.

File: packages/marketplace/src/db/queries.ts [hunk 7]
```diff
@@ -199,6 +203,190 @@ export async function getPluginVersion(
...
+		.bind(sanitizeReviewText(reply), reviewId)
+		.run();
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/db/queries.ts:377]: Valid. addPublisherReply() now validates that the sanitized reply is non-empty and returns null (404) if it would be blank. Fixed in 82915d0.

File: packages/marketplace/src/db/queries.ts [hunk 8]
```diff
@@ -97,6 +98,9 @@ export async function searchPlugins(
 		case "updated":
 			orderBy = "p.updated_at DESC";
 			break;
+		case "rating":
+			orderBy = "p.rating_avg DESC, p.rating_count DESC, p.created_at DESC";
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/db/queries.ts]: Valid. Rating sort now uses CASE WHEN rating_count >= 3 THEN rating_avg END DESC, so plugins below the display threshold sort to the bottom. Fixed in 82915d0.

File: packages/marketplace/src/db/schema.sql [hunk 9]
```diff
@@ -86,6 +88,25 @@ CREATE TABLE IF NOT EXISTS installs (
...
+  author_id TEXT NOT NULL REFERENCES authors(id),
+  rating INTEGER NOT NULL CHECK(rating BETWEEN 1 AND 5),
+  body TEXT,
+  publisher_reply TEXT,
+  replied_at TEXT,
+  created_at TEXT NOT NULL DEFAULT (datetime('now')),
+  updated_at TEXT NOT NULL DEFAULT (datetime('now')),
+  UNIQUE(plugin_id, author_id)
+);
+CREATE INDEX IF NOT EXISTS idx_reviews_plugin ON reviews(plugin_id);
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/src/db/schema.sql]: Valid. Changed to composite index on (plugin_id, created_at DESC) matching the query pattern. Fixed in 82915d0.

File: packages/marketplace/tests/reviews.test.ts [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+describe("review CRUD", () => {
+	let d1: ReturnType<typeof createD1Mock>;
+	let env: Record<string, unknown>;
+	let reviewerToken: string;
+
+	beforeEach(async () => {
+		resetRateLimiter();
+		d1 = createD1Mock();
+		env = makeEnv(d1);
+
```
[the remark this answers is not in the document]
Comment (BenjaminPrice) [reply packages/marketplace/tests/reviews.test.ts:140]: Valid. Added a test that creates a review, then attempts a second on a different plugin within the rate limit window and asserts 429. Fixed in 82915d0.

**Comment by ascorbic** [comment]
Closing -- we're rethinking the marketplace approach (see #296 for the new direction). Reviews and ratings are part of the broader marketplace plan, but we'd rather decide overall shape in #296 before layering on features.

Thanks for the work. Please feel free to re-raise the reviews/ratings design in #296 so it's part of the new direction.
